pkg/storage/iterator.go

```diff
@@ -403,8 +406,14 @@ func fetchLazyChunks(ctx context.Context, chunks []*chunkenc.LazyChunk) error {
 			}
 			chks, err := fetcher.FetchChunks(ctx, chks, keys)
 			if err != nil {
+				if err.Error() == chunkenc.ErrInvalidChecksum.Error() || strings.Contains(err.Error(), chunkenc.ErrInvalidChecksum.Error()) {
```

Comment:
Just checking `err == chunkenc.ErrInvalidChecksum` is not working. We have to do `err.Error() == chunkenc.ErrInvalidChecksum.Error()`. 

Also, the error might come from cortex from here https://github.com/cortexproject/cortex/blob/02af2474da0cef0cb4076106e03528eed8830166/pkg/chunk/chunk.go#L263. So, we have to check if `stack` contains `invalid chunk checksum` error.

Reply:
We should use the err not the stringer, as the stringer function could change.

If it can also come from cortex you can also check that since they are exposing it.

Reply:
There is a catch here. Cortex returns errors with different types. Here it is `promql` https://github.com/cortexproject/cortex/blob/master/pkg/chunk/chunk_store_utils.go#L165.

So, comparing `err == chunkenc.ErrInvalidChecksum` is not possible since `chunkenc.ErrInvalidChecksum` is of type `errors.Error` and `err` will be of type `promql.ErrStorage`

Reply:
Totally understand I think you should make a function.

`func isInvalidCheckSum(err error) bool`

You could check if it is a `promql.ErrStorage` first, if so unwrap the error, then check it.

Bonus point this means this function can be easily testable and you should send it chunkenc.ErrInvalidChecksum from cortex and from Loki and wrapped with promql.ErrStorage and with a random error.

Reply:
Perfect. Will add this